### DISCUSSION TITLE
bib文件中 citationkey 包含引号时，所有参考文献无法正常编译

### DISCUSSION THREAD

**Discussion by dreistein543** [nju-lug/NJUThesis#135] [Q&A] [answered]
报错信息形如 `Citation 'xxxxx' undefined on input line # `

问题可能来源 以及 自己尝试解决但**无用**的办法：
- 编译次数不足 —— ❌
> 在编辑器中分别尝试 xelatex biber xelatex *2 以及 latexmk -xelatex
> 在 terminal 中分别尝试上述两种编译方式

- 文献引用格式不对 —— ❌
> 通过编辑器自动填充 citationkey
> 手动复制粘贴 citationkey

- bib文件未指定 —— ❌
> 分别尝试了 `\addbibresource{}` 和在 bib={} 设定中指定位置均无效
> 更改 bib 文件的路径
> 分别使用绝对路径和相对路径

- bib文件格式有误 —— ❓
> 更换 bib 文件 ✅，
> 使用同 bib 文件某些子集 ✅，某些子集❌
> >相同 bib 文件在别的模板中**可以使用**(同样指定用biblatex-biber方式编译文献且整体编译过程与南大论文模板完全一致)
🌟似乎只要 bib 文件子集的 **citationkey 包含引号**，就无法正常输出参考文献🌟

附上 bib 格式，以供测试
```bibtex
% 带单引号的 citationkey
@article{abulof2017'Can'tBuy,
  title = {`Can't Buy Me Legitimacy': The Elusive Stability of Mideast Rentier Regimes},
  author = {Abulof, Uriel},
  year = {2017},
  journal = {Journal of International Relations and Development},
  volume = {20},
  number = {1},
  pages = {55--79},
  doi = {10.1057/jird.2014.32},
  langid = {english}
}

% 带双引号的 citationkey
@article{dingjianping2017"Yi,
  title = {``一带一路''中的宗教风险研究},
  author = {{丁剑平} and {方琛琳}},
  year = {2017},
  journal = {财经研究},
  volume = {43},
  number = {09},
  pages = {134--145},
  langid = {cn}
}

% 不带引号的 citationkey
@article{adams2011Chiefexecutive,
  title = {Chief Executive Officer Incentives, Monitoring, and Corporate Risk Management: Evidence from Insurance Use},
  author = {Adams, Mike and Lin, Chen and Zou, Hong},
  year = {2011},
  journal = {Journal of risk and Insurance},
  volume = {78},
  number = {3},
  pages = {551--582},
  langid = {english}
}

```

**Comment by atxy-blip** [comment, marked as answer]
`biber` 的处理机制不允许 citation key 中包含引号。事实上引号被用作分隔符，可以试试将条目内所有的花括号换成引号，也是可以编译的。见 <https://tex.stackexchange.com/questions/96454/using-bibtex-keys-containing-parentheses-with-biber/96918#96918>

使用文献工具导出时忽略一下非法字符就好，譬如 JabRef 的设置项有：
![image](https://user-images.githubusercontent.com/63113546/164960788-2733dccc-7207-434b-841e-d75479d74b15.png)

**Reply by dreistein543** [reply]
zotero 对于删除字符自定义性不足，可以通过 :clean 来回避危险的符号
不影响的符号有英文冒号 : 和短横线 -

**Reply by zepinglee** [reply]
> zotero 对于删除字符自定义性不足，可以通过 :clean 来回避危险的符号 不影响的符号有英文冒号 : 和短横线 -

你是用 betterbibtex 插件导出的吗？

**Reply by dreistein543** [reply]
> 你是用 betterbibtex 插件导出的吗？

是的，通过 BBT 导出，另外，我刚刚发现了一种可以完全回避标点符号的方法——加入 `:nopunct` 的方式

总体而言，我现在使用如下 citation key 生成规则，可以得到相对较好的结果：
`[auth:lower:clean][year][Title:clean:nopunct:split-ideographs:skipwords:select=1,2]`
作者姓第一个单词小写+发表年份+标题前两个单词/字符拼音
使用 `:clean` 可以用于回避带声调的字母，将中文标点转为英文标点
使用 `:split-ideographs` 将字符拆解开，否则汉字标题会完全显示
使用 `:skipwords` 可以回避在标题前两个单词中出现类似 a the on 等无意义词

**Reply by zepinglee** [reply]
> 总体而言，我现在使用如下 citation key 生成规则，可以得到相对较好的结果： `[auth:lower:clean][year][Title:clean:nopunct:split-ideographs:skipwords:select=1,2]` 作者姓第一个单词小写+发表年份+标题前两个单词/字符拼音 使用 `:clean` 可以用于回避带声调的字母，将中文标点转为英文标点 使用 `:split-ideographs` 将字符拆解开，否则汉字标题会完全显示 使用 `:skipwords` 可以回避在标题前两个单词中出现类似 a the on 等无意义词

学习了，感谢分享！
